p2pserver/dht/routing_table.go

```diff
@@ -0,0 +1,281 @@
+					EvtType: types.Add,
+					Event:   node,
+				}
+				this.feedCh <- feed
+			}
+			return true
+		}
+	}
+
+	// Todo: if the bucket is full, use LRU to replace
+	if len(bucket.entries) >= types.BUCKET_SIZE {
+		// bucket is full
+		return false
+	}
+
+	bucket.entries = append(bucket.entries, node)
+
+	copy(bucket.entries[1:], bucket.entries[:])
+	bucket.entries[0] = node
+	if this.feedCh != nil {
+		feed := &types.FeedEvent{
+			EvtType: types.Add,
+			Event:   node,
+		}
+		this.feedCh <- feed
+	}
+
+	return true
+}
+
+// removeNode removes a node with a given id
+func (this *routingTable) removeNode(id types.NodeID) {
+	this.mu.Lock()
+	defer this.mu.Unlock()
+	_, bucket := this.locateBucket(id)
+
+	entries := bucket.entries[:0]
+	var node *types.Node
+	for _, entry := range bucket.entries {
+		if entry.ID != id {
+			entries = append(entries, entry)
+		} else {
+			node = entry
+		}
+	}
+	bucket.entries = entries
+
+	if node != nil && this.feedCh != nil {
+		feed := &types.FeedEvent{
+			EvtType: types.Del,
+			Event:   node,
+		}
+		this.feedCh <- feed
+	}
+}
+
+// getClosestNodes returns the num nodes in the table that are closest to a given id
+func (this *routingTable) getClosestNodes(num int, targetID types.NodeID) []*types.Node {
+	this.mu.RLock()
+	defer this.mu.RUnlock()
+	closestList := make([]*types.Node, 0, num)
+
+	index, _ := this.locateBucket(targetID)
+	buckets := []int{index}
+	i := index - 1
+	j := index + 1
+
+	for len(buckets) < types.BUCKET_NUM {
+		if j < types.BUCKET_NUM {
+			buckets = append(buckets, j)
+		}
+		if i >= 0 {
+			buckets = append(buckets, i)
+		}
+		i--
+		j++
+	}
+
+	for _, index := range buckets {
+		for _, entry := range this.buckets[index].entries {
+			closestList = append(closestList, entry)
+			if len(closestList) >= num {
+				return closestList
+			}
+		}
+	}
+	return closestList
+}
+
+// getTotalNodeNumInBukcet returns the number of node in a specified bucket
+func (this *routingTable) getTotalNodeNumInBukcet(bucket int) int {
+	this.mu.RLock()
+	defer this.mu.RUnlock()
+	b := this.buckets[bucket]
+	if b == nil {
+		return 0
+	}
+
+	return len(b.entries)
+}
+
+// getLastNodeInBucket returns last node in a specified bucket
+func (this *routingTable) getLastNodeInBucket(bucket int) *types.Node {
+	this.mu.RLock()
+	defer this.mu.RUnlock()
+	b := this.buckets[bucket]
+	if b == nil || len(b.entries) == 0 {
+		return nil
+	}
+
+	return b.entries[len(b.entries)-1]
+}
+
+// getDistance returns the distance between nodes
+func (this *routingTable) getDistance(id1, id2 types.NodeID) int {
+	dist := logdist(id1, id2)
+	return dist
+}
+
+func (this *routingTable) totalNodes() int {
+	this.mu.RLock()
+	defer this.mu.RUnlock()
+	var num int
+	for _, bucket := range this.buckets {
+		num += len(bucket.entries)
+	}
+	return num
+}
+
+// isNodeInBucket checks whether a given node is in the specified bucket
+func (this *routingTable) isNodeInBucket(id types.NodeID, bucket int) (*types.Node, bool) {
```

Comment:
this function is redundant .  the following code is the same as `node := this.routingTable.queryNode(targetID)`
```golang
	bucket, _ := this.routingTable.locateBucket(targetID)
	node, ret := this.routingTable.isNodeInBucket(targetID, bucket)
```